### PR TITLE
8314483: Optionally override copyright header in generated source

### PR DESCRIPTION
**Pull request by erikj79**
In the JDK build we have various build tools that generate source code from data files. For most of these tools, the source files are based on template files, which already have copyright headers, but for some, the complete source file is generated by the tool, which is providing the copyright header programatically. For the latter, we would like to implement an override mechanism in each tool so that we can change the copyright header from a custom makefile.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314483](https://bugs.openjdk.org/browse/JDK-8314483): Optionally override copyright header in generated source (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [a53cf2f2](https://git.openjdk.org/jdk/pull/15346/files/a53cf2f2c4618788cd122f36155cfd65e7160dcf)
 * [Iris Clark](https://openjdk.org/census#iris) (@irisclark - **Reviewer**) ⚠️ Review applies to [a53cf2f2](https://git.openjdk.org/jdk/pull/15346/files/a53cf2f2c4618788cd122f36155cfd65e7160dcf)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15346/head:pull/15346` \
`$ git checkout pull/15346`

Update a local copy of the PR: \
`$ git checkout pull/15346` \
`$ git pull https://git.openjdk.org/jdk.git pull/15346/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15346`

View PR using the GUI difftool: \
`$ git pr show -t 15346`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15346.diff">https://git.openjdk.org/jdk/pull/15346.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15346#issuecomment-1684006772)

### CODE REVIEW

File: make/jdk/src/classes/build/tools/generatelsrequivmaps/EquivMapsGenerator.java [hunk 1]
```diff
@@ -53,21 +54,31 @@
 public class EquivMapsGenerator {
 
     public static void main(String[] args) throws Exception {
-        if (args.length != 3) {
+        int i = 0;
+        if (args.length == 5 && "-jdk-header-template".equals(args[i])) {
+            jdkHeaderTemplate = new String(Files.readAllBytes(Paths.get(args[++i])),
+                    StandardCharsets.UTF_8);
+            i++;
+        } else if (args.length != 3) {
```
Comment (dholmes-ora) [inline make/jdk/src/classes/build/tools/generatelsrequivmaps/EquivMapsGenerator.java:62]: This is not very robust for detecting malformed invocations. It won't even detect a simple typo in the argument name. I would suggest something more elaborate e.g.
```
int i  0;
boolean valid = true;
if (args.length != 5 && args.length !=3) {
  valid = false;
} else if (args.length == 5) {
  if ( ! "-jdk-header-template".equals(args[i])) {
    valid = false;
  } else {
    jdkHeaderTemplate = new String(Files.readAllBytes(Paths.get(args[++i])),
                    StandardCharsets.UTF_8);
     i++;
  }
}
if (!valid) {
  System.err.println(...);
   System.exit(1);
}
String lsrFile = args[i++];
...
```

**Review by dholmes-ora — APPROVED** [review]
Seems reasonable.

One suggested change to make the tool command-line processing more robust, but up to you whether to take it or not.

Thanks.

File: make/jdk/src/classes/build/tools/generatelsrequivmaps/EquivMapsGenerator.java (hunk 1 above)
Comment (erikj79) [reply make/jdk/src/classes/build/tools/generatelsrequivmaps/EquivMapsGenerator.java]: I think it actually will, but your suggestion is clearer to read, so I have applied it.

**Review by irisclark — APPROVED** [review]

**Review by dholmes-ora — APPROVED** [review]
Looks good. Thanks

File: make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java [hunk 2]
```diff
@@ -229,6 +231,12 @@ public static void main(String[] args) throws Exception {
                         tzDataDir = args[++i];
                         break;
 
+                    case "-jdk-header-template":
+                        jdkHeaderTemplate = new String(
+                            Files.readAllBytes(Paths.get(args[++i])),
```
Comment (turbanoff) [inline make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java:236]: Can `java.nio.file.Files#readString(java.nio.file.Path)` be used instead?

File: make/jdk/src/classes/build/tools/generatelsrequivmaps/EquivMapsGenerator.java [hunk 3]
```diff
@@ -53,21 +54,39 @@
...
 
     public static void main(String[] args) throws Exception {
-        if (args.length != 3) {
+        int i = 0;
+        boolean valid = true;
+        if (args.length != 5 && args.length !=3) {
+            valid = false;
+        } else if (args.length == 5) {
+            if ("-jdk-header-template".equals(args[i])) {
+                jdkHeaderTemplate = new String(Files.readAllBytes(Paths.get(args[++i])),
```
Comment (turbanoff) [inline make/jdk/src/classes/build/tools/generatelsrequivmaps/EquivMapsGenerator.java:63]: Can be simplified with `Files.readString(java.nio.file.Path)`

File: make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java (hunk 2 above)
Comment (erikj79) [reply make/jdk/src/classes/build/tools/cldrconverter/CLDRConverter.java]: Ah yes, I was looking for that method and didn't find it, but I now realize I was looking in the jdk 8 api. Punished for not setting up an IDE for editing these files. Fixed now, thanks!

**Pull request closed without merging**